Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2ClientRequestHandler.swift

```diff
@@ -0,0 +1,242 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import NIO
+import NIOHTTP1
+import NIOHTTP2
+
+class HTTP2ClientRequestHandler: ChannelDuplexHandler {
+    typealias OutboundIn = HTTPExecutingRequest
+    typealias OutboundOut = HTTPClientRequestPart
+    typealias InboundIn = HTTPClientResponsePart
+
+    private let eventLoop: EventLoop
+
+    private(set) var channelContext: ChannelHandlerContext?
+    private(set) var state: HTTPRequestStateMachine = .init(isChannelWritable: false)
+    private(set) var request: HTTPExecutingRequest?
+
+    init(eventLoop: EventLoop) {
+        self.eventLoop = eventLoop
+    }
+
+    func handlerAdded(context: ChannelHandlerContext) {
+        self.channelContext = context
+
+        let isWritable = context.channel.isActive && context.channel.isWritable
+        let action = self.state.writabilityChanged(writable: isWritable)
```

Comment:
Will the state machine tolerate this event before the first `write`?

Reply:
Added a test case to the `HTTPRequestStateMachineTests`: `testConnectionBecomesWritableBeforeFirstRequest`.